lib/runtime/formatter.js

```diff
@@ -66,3 +67,35 @@ function markRange(editor, range) {
   })
   return marker
 }
+
+let subs
+
+export function activate() {
+  subs = new CompositeDisposable()
+  edWatch = new WeakSet()
+
+  subs.add(atom.workspace.observeTextEditors(ed => {
+    edWatch.add(ed)
+    let edsub = ed.getBuffer().onDidSave(() => {
+      if (ed && ed.getGrammar && ed.getGrammar().id === 'source.julia') {
+        if (client.isActive() && edWatch.has(ed)) {
+          formatEditor(ed).then(() => {
+            edWatch.delete(ed)
+            ed.save().then(() => {
+              edWatch.add(ed)
+            })
+          })
+        }
+      }
+    })
+    subs.add(edsub)
```

Comment:
I guess we can simplify the logic by using [`TextBuffer.onWillSave`](https://flight-manual.atom.io/api/v1.43.0/TextBuffer/#instance-onWillSave) (and removing `edWatch`) ?

Reply:
No, because the order in which `onWillSave` hooks run isn't really specified (technically it is based on package activation order, I think). 
If we were to just use `onWillSave` then we'd grab a certain buffer state and invoke Julia's formatter, but in the meantime the whitespace package might modify the code, which means that our formatter fails (hence the `force` option I had above).